tensorboard/uploader/formatters.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Helpers that format experiment metadata as strings."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import json
+
+EXPERIMENT_METADATA_URL_JSON_KEY = "url"
+ExperimentMetadataField = collections.namedtuple(
+    "ExperimentMetadataField",
+    ("json_key", "readable_name", "value", "formatter"),
```

Comment:
Adding a formatter interface sounds reasonable to me, but this doesn’t
really feel like the right structure. The `ExperimentMetadataField`
struct is tightly coupled to the two implementations of formatters. The
`readable_name` nad `formatter` fields are only relevant to the
`ReadableFormatter`, and the `json_key` field is only relevant to the
`JsonFormatter`. And as written, although the name `Experiment*` is
everywhere, nothing actually knows about experiments.

Why introduce a new type with a big list of vague “fields” when we
already have the experiment proto? The interface could just be

```python
class ExperimentFormatter(metaclass=abc.ABCMeta):
    def format(self, experiment):
        """Format an experiment.

        Args:
          experiment: An `experiment_pb2.Experiment` value.

        Returns:
          A string.
        """
        pass
```

The `ReadableExperimentFormatter` would encapsulate the readable field
names and how to format them, and the `JsonExperiment` would encapsulate
the JSON keys and how to order and indent them. The tests would be on
the same output that we would actually see from the binary rather than
just an arbitrary set of fields defined in the test case.

Then, you could remove the giant block of field definitions in
`uploader_main.py`, too: just `print(formatter.format(experiment))`.

Reply:
Done effecting the refactoring that you suggested. This revision has pushed a lot of the variables related to the formatting into `formatters.py`, which is indeed nicer than before. 

Also note that I changed the signature of the `format_experiment()` method slightly so that it takes two input arguments: `experiment` (a `experiment_pb2.Experiment` proto) and `experiment_url` (a string). Getting the latter from an `Experiment` proto requires the `server_info` module, which I think shouldn't be known to the formatter module, which is purely concerned with string formatting.